tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -502,25 +529,38 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
       _cache
     }
 
-    def getStaticCache(encodedName: String): MethodCache =
+    def getStaticCache(encodedName: String): MethodCache[js.Tree] =
       _staticCaches.getOrElseUpdate(encodedName, new MethodCache)
 
-    def getMethodCache(encodedName: String): MethodCache =
+    def getMethodCache(encodedName: String): MethodCache[js.Tree] =
       _methodCaches.getOrElseUpdate(encodedName, new MethodCache)
 
+    def getConstructorCache(): MethodCache[js.Tree] = {
+      _constructorCache.getOrElse {
+        val cache = new MethodCache[js.Tree]
+        _constructorCache = Some(cache)
+        cache
+      }
+    }
+
     def cleanAfterRun(): Boolean = {
       _staticCaches.retain((_, c) => c.cleanAfterRun())
       _methodCaches.retain((_, c) => c.cleanAfterRun())
 
+      if (_constructorCache.exists(!_.cleanAfterRun()))
+        _constructorCache = None
+
       if (!_cacheUsed)
         invalidate()
 
       _staticCaches.nonEmpty || _methodCaches.nonEmpty || _cacheUsed
     }
   }
 
-  private final class MethodCache extends knowledgeGuardian.KnowledgeAccessor {
-    private[this] var _tree: WithGlobals[js.Tree] = null
+  private final class MethodCache[TreeType]
+      extends knowledgeGuardian.KnowledgeAccessor {
```

Comment:
I can actually revert that generalization with a type parameter. Now all uses of `MethodCache` are back to using `js.Tree`.